regression/contracts/quantifiers-exists-ensures-02/main.c

```diff
@@ -1,20 +1,41 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#define MAX_LEN 128
+
 // clang-format off
-int f1(int *arr) __CPROVER_ensures(__CPROVER_exists {
-  int i;
-  (0 <= i && i < 10) ==> arr[i] != 0
-})
+int f1(int *arr, int len)
+  __CPROVER_ensures(
+    len > 0 ==> __CPROVER_exists {
+      int i;
+      // test replacement with symbolic bound
+      (0 <= i && i < len) && arr[i] == 0
+    }
+  )
 // clang-format on
 {
-  for(int i = 0; i < 10; i++)
-  {
-    arr[i] = 0;
-  }
-
+  // we are only checking for contract replacement
   return 0;
 }
 
 int main()
 {
-  int arr[10];
-  f1(arr);
+  int len;
+  __CPROVER_assume(0 <= len && len <= MAX_LEN);
+
+  int *arr = malloc(len * sizeof(int));
+
+  f1(arr, len);
+
+  bool found_zero = false;
+  for(int i = 0; i <= MAX_LEN; i++)
+  {
+    if(i < len)
+      found_zero |= (arr[i] == 0);
+  }
```

Comment:
We are not proving these programs are correct, because we're not unrolling these loops. Thus, we have two options here:
1. Add invariants to these loops;
2. Update our regression-test script to also call CBMC with the correct bounds and `--unwinding-assertions`;

I'd prefer option (1). In any case, we can create a GitHub issue for this now and tackle it on a separate PR (for all test cases).

Reply:
But all loops in CBMC are always unrolled. This one is too, you can see in the CBMC output.

With symbolic bounds the unrolling just never stops, so we give `--unwind` to limit unrolling and then `--unwinding-assertions` to make sure our limited unrolling is sound.

Another way to stop the unrolling is to provide an invariant, and then we just reason about a single arbitrary iteration.

Here, since the loop bounds are constants (0 and 128), the unrolling stops automatically, and is always sound since there is nothing more to unroll (i.e. we didn't limit anything).